File: shared/types/MessageDTO.ts
// interface til en melding for å matche DTO fra backend. Inneholder AttachmentDTO, ReactionDTO, MessageDTO og SendMessageRequestDTO
import { UserSummaryDTO } from "./UserSummaryDTO";

export interface AttachmentDto {
  fileUrl: string;
  fileType: string;
  fileName: string;

  isOptimistic?: boolean;
  optimisticId?: string; // Temporary ID før vi får riktig ID fra backend
  localUri?: string; // Local file URI for preview (React Native)
  isUploading?: boolean;
  uploadError?: string | null;
  fileSize?: number; 
  isEncrypted?: boolean;
  needsDecryption?: boolean;

   // Encryption metadata for lazy decryption
  keyInfo?: { [userId: string]: string };
  iv?: string;
  version?: number;

  thumbnailUrl?: string;           // Encrypted thumbnail URL from server
  thumbnailWidth?: number;         // Original thumbnail dimensions
  thumbnailHeight?: number;
  thumbnailKeyInfo?: { [userId: string]: string }; // Encryption keys for thumbnail
  thumbnailIV?: string;            // Thumbnail encryption IV
  localThumbnailUri?: string;   
}

export interface ReactionDTO {
  messageId: number;
  emoji: string;
  userId: number;
  isRemoved: boolean;
  userFullName?: string;
  conversationId: number; 
}

export interface MessageDTO {
  id: number;
  senderId: number | null;
  text: string | null;
  sentAt: string; // ISO-dato som kommer fra backend
  conversationId: number;
  attachments: AttachmentDto[];
  reactions: ReactionDTO[];
  parentMessageId?: number | null;
  parentMessageText?: string | null;
  parentSender?: UserSummaryDTO | null;
  isRejectedRequest?: boolean;
  sender?: UserSummaryDTO | null;
  isNowApproved?: boolean;
  isSilent?: boolean;
  isSystemMessage: boolean;
  isDeleted?: boolean;

  // Optimistic fields - frontend only
  isOptimistic?: boolean;
  optimisticId?: string; // Temporary ID før vi får riktig ID fra backend
  isSending?: boolean;
  sendError?: string | null;
}


export interface SendMessageRequestDTO {
  text?: string;
  attachments?: AttachmentDto[];
  conversationId?: number;
  receiverId?: string;
  parentMessageId?: number | null;
}

export interface UploadAttachmentsRequestDTO {
  text?: string;
  files: File[];
  conversationId: number;
  receiverId?: string;
  parentMessageId?: number | null;
}

// Hjelpetype for å bygge FormData
export interface MessageWithFilesData {
  text?: string;
  files?: File[];
  conversationId: number;
  receiverId?: string;
  parentMessageId?: number | null;
}

File: AFBack/Features/Messaging/DTOs/MessageDto.cs
using AFBack.Features.Reactions.DTOs;
using AFBack.Features.Reactions.DTOs.Responses;

namespace AFBack.Features.Messaging.DTOs;

/// <summary>
/// Intern DTO for message-data fra databasen. Brukes som mellomformat mellom Repository og Service-lag.
/// MessageResponse er det frontend forventer
/// </summary>
public class MessageDto
{
    public int Id { get; set; }
    public string? SenderId { get; set; }
    public string? EncryptedText { get; set; }
    public string KeyInfo { get; set; } = null!;
    public string IV { get; set; } = null!;
    public int Version { get; set; }
    public DateTime SentAt { get; set; }
    public int ConversationId { get; set; }
    public bool IsDeleted { get; set; }
    public int? ParentMessageId { get; set; }
    public string? ParentMessagePreview { get; set; }
    public bool IsSystemMessage { get; set; }
    public string? ParentSenderId { get; set; }
    
    public List<AttachmentDto> Attachments { get; set; } = new();
    public List<ReactionResponse> Reactions { get; set; } = new();
}
